map/map_tests/bookmarks_test.cpp

```diff
@@ -1387,7 +1388,82 @@ UNIT_CLASS_TEST(Runner, Bookmarks_AutoSave)
+  bmManager.EnableTestMode(true);
+
+  BookmarkManager::KMLDataCollection kmlDataCollection;
+  for (auto const & file : files)
+    kmlDataCollection.emplace_back(file, LoadKmlFile(file, KmlFileType::Gpx));
+
+  bmManager.CreateCategories(std::move(kmlDataCollection));
+  TEST_EQUAL(bmManager.GetBmGroupsIdList().size(), 4, ());
+
+  auto categories = bmManager.GetBmGroupsIdList();
+  auto checker = [](BookmarkManager::SharingResult const & result)
+  {
+    auto kmz = result.m_sharingPath;
+    ZipFileReader::FileList files;
+    ZipFileReader::FilesList(kmz, files);
+    TEST_EQUAL(files.size(), 5, ("5 files are expected in kmz"));
+    auto index = "doc.kml";
+    std::vector<std::string> expectedFiles = {"doc.kml", "files/new.kml", "files/Some random route.kml",
+                                               "files/OrganicMaps_1.kml", "files/OrganicMaps_2.kml"};
+    for (auto const & file : files)
+      TEST(std::find(expectedFiles.begin(), expectedFiles.end(), file.first) != expectedFiles.end(), ());
+    auto indexPath = base::JoinPath(GetPlatform().TmpDir(), index);
+    ZipFileReader::UnzipFile(kmz, index, indexPath);
+    std::string indexContent;
+    FileReader(indexPath).ReadAsString(indexContent);
+    std::string expectedIndexContent;
+    FileReader(GetPlatform().TestsDataPathForFile("kml_test_data/kmz_index.kml"))
+        .ReadAsString(expectedIndexContent);
+    TEST_EQUAL(expectedIndexContent, indexContent, ("Index content doesnt match expected value"));
+    auto tmpPath = base::JoinPath(GetPlatform().TmpDir(), "tmp.xml");
+    for (auto const & file : files)
+    {
+      base::DeleteFileX(tmpPath);
+      ZipFileReader::UnzipFile(kmz, file.first, tmpPath);
+    }
+    TEST(base::DeleteFileX(kmz), ());
+    TEST(base::DeleteFileX(indexPath), ());
+    TEST(base::DeleteFileX(tmpPath), ());
+  };
+  bmManager.PrepareFileForSharing(std::move(categories), checker);
+}
+
+UNIT_CLASS_TEST(Runner, ExportSingleUnicode)
+{
+  string file = GetPlatform().TestsDataPathForFile("gpx_test_data/Üφήが1.gpx");
+  BookmarkManager bmManager(BM_CALLBACKS);
+  bmManager.EnableTestMode(true);
+  BookmarkManager::KMLDataCollection kmlDataCollection;
+  kmlDataCollection.emplace_back(file, LoadKmlFile(file, KmlFileType::Gpx));
+  bmManager.CreateCategories(std::move(kmlDataCollection));
+  auto categories = bmManager.GetBmGroupsIdList();
+  auto checker = [](BookmarkManager::SharingResult const & result)
+  {
+    auto kmz = result.m_sharingPath;
+    ZipFileReader::FileList files;
+    ZipFileReader::FilesList(kmz, files);
+    TEST_EQUAL(1, files.size(), ());
+    TEST_EQUAL("OrganicMaps.kml", files.at(0).first, ());
+    auto tmpPath = base::JoinPath(GetPlatform().TmpDir(), "tmp.xml");
+    ZipFileReader::UnzipFile(kmz, files.at(0).first, tmpPath);
+    TEST(base::DeleteFileX(kmz), ());
+    TEST(base::DeleteFileX(tmpPath), ());
+  };
+  bmManager.PrepareFileForSharing(std::move(categories), checker);
```

Comment:
The checker's code is run here properly because of the test mode, right?

Reply:
Yes, as I understand it works because we submit handler to UI thread from main thread of the test (that is a UI thread also?). If we use standard flow of bookmark manager (without locking) we have test_thread -> file_thread -> submit_to_ui_thread - and handler is not executed because tests ends earlier. Current test flow looks a bit fragile for me, I had an idea maybe to add something like
```

    TEST(base::DeleteFileX(indexPath), ());
    TEST(base::DeleteFileX(tmpPath), ());
    **checked = true;**
  };
  bmManager.PrepareFileForSharing(std::move(categories), checker);
  **TEST(checked, ());**
```
to verify that checker was called, but not sure - maybe it is a boilerplate.